controller_manager/controller_manager/test_utils.py

```diff
@@ -0,0 +1,129 @@
+#      notice, this list of conditions and the following disclaimer in the
+#      documentation and/or other materials provided with the distribution.
+#
+#    * Neither the name of the {copyright_holder} nor the names of its
+#      contributors may be used to endorse or promote products derived from
+#      this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+# Author: Christoph Froehlich
+
+import time
+
+from launch_testing_ros import WaitForTopics
+from sensor_msgs.msg import JointState
+from controller_manager.controller_manager_services import list_controllers
+
+
+def check_node_running(node, node_name, timeout=5.0):
+
+    start = time.time()
+    found = False
+    while time.time() - start < timeout and not found:
+        found = node_name in node.get_node_names()
+        time.sleep(0.1)
+    assert found, f"{node_name} not found!"
+
+
+def check_controllers_running(node, cnames, namespace="", state="active"):
+    """
+    Check if the specified controllers are running on the given node.
+
+    Args:
+      node (Node): The ROS2 node instance to check for controllers.
+      cnames (list of str): List of controller names to check.
+      namespace (str, optional): The namespace in which to look for controllers. Defaults to "".
+      state (str, optional): The desired state of the controllers. Defaults to "active".
+
+    Raises:
+      AssertionError: If any of the specified controllers are not found or not in the desired state within the timeout period.
+    """
+
+    # wait for controller to be loaded before we call the CM services
+    found = {cname: False for cname in cnames}  # Define 'found' as a dictionary
+    start = time.time()
+    # namespace is either "/" (empty) or "/ns" if set
+    if namespace:
+        namespace_api = namespace
+        if not namespace_api.startswith("/"):
+            namespace_api = "/" + namespace_api
+        if namespace.endswith("/"):
+            namespace_api = namespace_api[:-1]
+    else:
+        namespace_api = "/"
+
+    while time.time() - start < 10.0 and not all(found.values()):
+        node_names_namespaces = node.get_node_names_and_namespaces()
```

Comment:
Wasn't that the functionality that was causing trouble for us in the spawner when waiting for the controller manager?

Reply:
I'm not sure, I can't remember. I added this some time ago in the demo tests, and it seemed to be useful back then ;) maybe the list_controllers service is now more robust and we could skip this node check in advance.

Reply:
Atleast in the ros2_control_demos, we haven't seen flaky tests due to this right?. We can remove it, if this is not interesting for us to have consistency or we can get back to it when we see some flakiness. Both are fine for me

Reply:
no it didn't make any problems recently

Reply:
I'll approve it as I don't have a strong opinion on this